### PR TITLE
Step Completion

### PR DESCRIPTION
**Pull request by Socolin**
See documentation in [doc/how-it-works/completion.md](https://github.com/SpecFlowOSS/SpecFlow.Rider/blob/completion/doc/how-it-works/completion.md)

Reviews are welcome.
I want to add some tests too, there are a lot of edge case here.

### CODE REVIEW

File: src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/Caching/StepsDefinitions/AssemblyStepDefinitions/SpecflowStepInfoFactory.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                switch (type)
+                {
+                    case StepPatternUtil.StepPatternTokenType.Text:
+                        partialPattern.Append(text);
+                        break;
+                    case StepPatternUtil.StepPatternTokenType.Capture:
+                        var captureText = text;
+                        if (text == ".+")
+                            captureText = ".+?";
+                        else ;if (text == ".*")
```
Comment (nemesv) [inline src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/Caching/StepsDefinitions/AssemblyStepDefinitions/SpecflowStepInfoFactory.cs:68]: This seems to be a wrongly placed semicolon in `else ;if`

File: src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinStepCompletionProvider.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+            var psiModule = context.BasicContext.File.GetPsiModule();
+            var selectedStepKind = selectedStep.GetStepKind();
+            foreach (var stepDefinitionInfo in specflowStepsDefinitionsCache.GetStepAccessibleForModule(psiModule, selectedStepKind).Concat(assemblyStepDefinitionCache.GetStepAccessibleForModule(psiModule, selectedStepKind)))
+            {
+                if (stepDefinitionInfo.RegexForPartialMatch == null)
+                    continue;
+
+                var partialStepText = selectedStep.GetStepTextBeforeCaret(context.BasicContext.CaretDocumentOffset);
+                var fullStepText = selectedStep.GetStepText();
```
Comment (nemesv) [inline src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinStepCompletionProvider.cs:42]: Calculating the stepText can be moved outside the foreach loop

File: src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinKeywordCompletionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        protected override bool IsAvailable(GherkinSpecificCodeCompletionContext context)
+        {
+            var codeCompletionType = context.BasicContext.CodeCompletionType;
+            return codeCompletionType == CodeCompletionType.BasicCompletion || codeCompletionType == CodeCompletionType.SmartCompletion;
+        }
+
+        protected override bool AddLookupItems(GherkinSpecificCodeCompletionContext context, IItemsCollector collector)
+        {
+            return false;
```
Comment (nemesv) [inline src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinKeywordCompletionProvider.cs:20]: Do we want to include the KeyWord completation in this PR or it will be added later as a separate step?

**Comment by nemesv** [comment]
The documentation is very good and helpful to understand the feature. 👍

File: src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinKeywordCompletionProvider.cs (hunk 3 above)
Comment (Socolin) [reply src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/CompletionProviders/GherkinKeywordCompletionProvider.cs:20]: No, I wanted to start with this, but I hit a problem where the completion would not show up, so I asked in the slack channel for help and while I was waiting I started the step completion, and I finally fix the problem while doing it. I'll remove it to avoid confusion

**Comment by nemesv** [comment]
@Socolin it seems with the namespace rename something has changed in the build order, and the generation of the CSharpErrors.generated.cs is executed later which fails the build.

Locally in Rider it is always working, manually executing the buildPlugin gradle task is working mostly... 
But it was also failed 2 times, however I cannot really reproduce it anymore. 

What I've tried to delete `CSharpErrors.generated.cs` and rerun the task, but in most cases it works. However in the build it always fails. I've tried to restart multiple times.

**Comment by Socolin** [comment]
Yes I do not understand what is the problem, I'll take a look the weekend

**Comment by Socolin** [comment]
So I had to optimize some stuff in the Regex library, it was taking 3 minutes to parse the regex in my work project (not that big) so it was not usable. Now it feel reactive and quick enough, this problem should be defenitively fixed once we can  use PCRE.NET for this.

After testing with my projects I also found some details, when a step contains a `\(` or `\)` etc.. so I Unescape the pattern before showing it in completion

Completion is also now handling optional group / characters correctly (when there are `?` in pattern)
